### ISSUE TITLE
[Question] How to get correct fio results

### ISSUE THREAD

**Issue by filipLoisel** [axboe/fio#1059]
I am a novice when it comes to testing I/O performance and have little experience. Currently, I have a client node that mounts `CephFS` into /mnt/storage/` and from there I'm creating fio requests to benchmark the performance.

I'm using following command on the client:
```bash
$ sudo fio --fallocate=none --name=cephfs --rw={read|randread|write|randwrite} --size=1M --bs=4K --directory=/mnt/storage --ioengine=sync --numjobs=1
```
The size varies from 4K up to 65,536K

With 1M, I notice a difference when executing fio the first time. E.g., with `--rw=read`, I receive for the first time run about 1800 IOPS. After that, always about 2500 IOPS.

For the write operations, it's the other way around. Fio reports for the first execution 21.3k(!) IOPS. When I'm not removing `/mnt/storage/cephfs.0.0`, it's about 873 IOPS.

I am quite confused and do not know which numbers to use for my report. Is that normal in fio, or is it a bug?

fio version: 3.1
client os: Ubuntu Server 18.04.4 for Raspberry Pi 3
cluster os: Raspberian Buster
The devices are Raspberry Pi 3 

I am very grateful for quick help!

**Comment by sitsofe** [comment]
@filipLoisel: The things I will note about your job:
- You are doing very little I/O (1 megabyte tops) which ties in with the point below
- Your I/O is allowed to be cached. Reads may be being satisfied by some cache if invalidation isn't happening and there's no guarantee that write I/O isn't just sitting in memory (and thus hasn't yet reached your Ceph backend) by the time the job completes
- You are doing I/O synchronously (no more than one I/O in flight at any one time)
- You've turned off preallocation so a job that contains writes alone may wind up with each write having to extended the file on every new write if the file doesn't already exist

> With 1M, I notice a difference when executing fio the first time. E.g., with --rw=read, I receive for the first time run about 1800 IOPS. After that, always about 2500 IOPS.

Hmm, you didn't include the fio summary output so we don't have any clues that would provided. I'd guess there's some caching going on? I suppose you could delete the file between runs (e.g. [`--unlink`](https://fio.readthedocs.io/en/latest/fio_man.html#cmdoption-arg-unlink)) and see what happens...

> For the write operations, it's the other way around. Fio reports for the first execution 21.3k(!) IOPS. When I'm not removing /mnt/storage/cephfs.0.0, it's about 873 IOPS.

I'd would have expected both runs to behave similarly but I don't know what CephFS is going to do. Do the results change if you delete the file and you use [`--end_fsync=1`](https://fio.readthedocs.io/en/latest/fio_man.html#cmdoption-arg-end-fsync) on all runs?

> I am quite confused and do not know which numbers to use for my report. Is that normal in fio, or is it a bug?

If you do only enough I/O that it can be satisfied by cache/buffering then seeing "odd" and extremely fast speeds is not unusual. Something you could try is making `size` at least twice the memory of your machine as that will mean much of the I/O won't be cacheable in the Linux page cache and will hopefully lead to more stable results. From what you've told us above I'd strongly suggest you contact the Ceph folks (https://ceph.io/irc/ but note that page links to their mailing lists too) because they are more likely to be able to explain your results and tell you if caching is likely etc. Based on what you've said so far it doesn't sound like a bug in fio.

In summary: I'd reach out to the Ceph community and see if they have suggestions. I'm going to close this issue here but if you find out the answer please post a follow up! You could also try asking on the [fio mailing list](https://github.com/axboe/fio/blob/fio-3.20/README#L58) (note that the list ONLY accepts plain text emails) as that is better for "Why is this happening?" style questions (whereas the github tracker is for definite bugs).

**Comment by filipLoisel** [comment]
Thank you very much for your help!. I have applied your tips and with the mentioned flag options, I already get much more consistent data
